modules/renter/contractor/journal.go

```diff
@@ -0,0 +1,492 @@
+package contractor
+
+// The contractor achieves efficient persistence using a JSON transaction
+// journal. It enables efficient ACID transactions on JSON objects.
+//
+// Each journal represents a single JSON object. The object is serialized as
+// an "initial object" followed by a series of update sets, one per line. Each
+// update specifies a field and a modification. See the journalUpdate type for
+// a full specification.
+//
+// During operation, the object is first loaded by reading the file and
+// applying each update to the initial object. It is subsequently modified by
+// appending update sets to the file, one per line. At any time, a
+// "checkpoint" may be created, which clears the journal and starts over with
+// a new initial object. This allows for compaction of the journal file.
+//
+// In the event of power failure or other serious disruption, the most recent
+// update set may be only partially written. Partially written update sets are
+// simply ignored when reading the journal. Individual updates may also be
+// ignored if they are malformed, though other updates in the set may be
+// applied. See the journalUpdate docstring for an explanation of malformed updates.
+
+// TODO:
+// - handle case sensitivity
+// - handle null vs. empty array
+//    - if index is 0, just do a wholesale replace with [val]
+// - handle inserting new keys into object
```

Comment:
I think switching to strict setters eliminates all of these issues. Not sure how jarring of a change that would be though.